### ISSUE TITLE
Cygwin - Windows 7 arrow keys don't work

### ISSUE THREAD

**Issue by mcormier** [fish-shell/fish-shell#1160]
I'm running Cygwin on a windows 7 machine.  When I run fish it works but the arrow keys do not.  My first guess was that this was a configuration issue, but I could not figure out what needed to be changed.

```
prompt )
           [C
           [D
           [C
           [D
           [C
           [A
           [B
```

Also, this is running on an "old-style" cygwin window where it is not a tty emulator, but similar to the dos window that you get when you run cmd.exe.

**Comment by knight-of-design** [comment]
Encountered the Arrow Key ([C) issue using the cygwin fish binary.  However, a workaround that works for me is starting fish within mintty.  

`/cygwin64/bin/mintty.exe -e '/bin/fish' `

**Comment by flip4bytes** [comment]
@knight-of-design I am encountering the exact same problem, same exact setup as you and everything. I've been digging down for the past 3 hours and have like 50 tabs open trying to figure out wtf is going on. I too noticed that starting fish within mintty allows it to work normally, though starting it any other way causes the `[A [B [C [D` problem.

**Comment by johnlindquist** [comment]
I have the same problem using fishshell inside of cmder or launching the fish.exe directly..

Oddly enough, arrow keys work fine using fishshell inside of WebStorm/IntelliJ's terminal emulation. Not sure if that helps anyone though.

**Comment by RichardClaxton** [comment]
C:\cygwin64\bin\mintty.exe -e '/bin/fish'

Good Call !

**Comment by theonewolf** [comment]
Hitting this same issue using ConEmu.

**Comment by alkemir** [comment]
Any lights on this issue? It is really a deal breaker, such a pity.

I compiled version 2.2b1-39-g4621e76 and noticed a change in behaviour: The newline is replaced by a ? sign, that is, pressing up now yields ?[A.

I'll stick around see if I can be of assistance.

**Comment by zanchey** [comment]
The newline problem is #2092.

I've been playing with terminal modes and termcap to see if I can work out what's going on. Strangely, arrow keys work to move the cursor under `cat`, which is a bit concerning as the terminal should be in raw mode.

**Comment by alkemir** [comment]
So, bind is not catching our arrow keys; because apparently the signal is different (is there a way to print scancodes, char codes or something like that?). But, if you do
`bind \[C forward-char`  you DO get a right-arrow behaviour. Of course the problem is that whatever non-printable chars that prefix the \[C don't get captured and mess with any attempt to use the arrows.

Issuing new binds would patch this for us, if we knew what the prefix is.
In my case I get "fish: Wide character 65535 has no narrow representation" , but I am guessing there is a better way to get the actual chars than this. A concern I am starting to humor is whether bind supports the use of these non-printable chars.

**Comment by ridiculousfish** [comment]
@alkemir You can try using `read` under `bash`to get the characters that are being sent.

**Comment by alkemir** [comment]
@ridiculousfish I used xxd -p, which should work as well.
Apparently this is a dead end, as the characters sent are 1b 5b 43;
1b is escape (\e) 5b is [ and 43 is C ; for some reason bind is not capturing
the escape char. This might have to do with the console program capturing it before?
Although, on second tought, then xxd wouldn't be able to catch it... I am at a loss here.

If you send me commands I can be your hands, but I am really new to all this.

**Comment by alkemir** [comment]
Just realized that console2 is behaving very strangely. If I order it to use
bash or fish as shell I shrinks the screen and becomes unusable. There I AM
using cmd.exe wrapped by Console2. Sorry for the misunderstanding.

I tried ansiCons and it didn't help, but I realized that when I press Ctrl + C I also get the
`fish: Wide character 65535 has no narrow representation` error; thus I might have
solved partially the keys problem with the bindings above, but I am encountering a new error.

To sum up: \c and \e seem broken in Windows using cmd.exe

**Comment by alkemir** [comment]
Ok, great news!

bind \cj backward-char 

works as expected (and provides an usable alternative, using ijkl)
But \ej doesn't get caught. There might be an issue with the meaning of \e in Windows?
Maybe it is (I have no idea what I am talking about) that \e and \c are being sent as wide chars?
Do you think this has anything to do with the platform being x64?

**Comment by alkemir** [comment]
I feel we are close!

I turned on debugging and amazing as it seems, fish is getting 0xffff instead of 0x1b . This explains both the failure to capture arrow keys (and others) and the "Wide character.. .." error message.

I am unsure where this happens, I will keep poking around but this should really bring us close the solving this issue.

**Comment by alkemir** [comment]
Great news!

Apparently this has something to do with timing; I am not very familiar with the code or keyboard handling at all, but I have a working patch!

Rewrite the for loop inside input_mapping_is_match (input.cpp:644) to look like this:
`    for (j=0; str[j] != L'\0'; j++)
    {
        bool timed = (j > 0 && iswcntrl(str[0]));

        c = input_common_readch(timed);
        if (c == 0xffff)
        {
            j--;
            continue;
        }
        if (str[j] != c)
        {
            break;
        }
    }`

It is an ugly patch waiting for the devs to figure out why they are sending 0xffff on timeout (maybe the timeout is badly detected, or 0xffff is WEOF or something like that on Win7), but this should be easy to figure out for somebody who knows the code. In the meantime, we can patch it and have fish working on Windows!

**Comment by ridiculousfish** [comment]
(I fixed your code formatting)

**Comment by flip4bytes** [comment]
woohoo! so @ridiculousfish does that mean that this is finally going to be officially solved in the next release? or is this just a temporary fix?

**Comment by zanchey** [comment]
I'll leave that to the Cygwin maintainer, @xfix, as I don't think we'll be taking this on the main branch for 2.2.0.

**Comment by ridiculousfish** [comment]
@alkemir Very interesting! If you increase `WAIT_ON_ESCAPE` to, say, 100, does that also fix the problem?

**Comment by alkemir** [comment]
@ridiculousfish Yes, setting `WAIT_ON_ESCAPE = 100` does fix the problem, with the only issue that if you leave the button pressed (like you would to to go back 10 characters) sometimes you get [D . I guess this is the source of the problem, the escape character sometimes arrives separated from the rest of the characters. Setting it to 1000 eliminates the spurious escapees; so I guess there is a sweetspot we must find, and that would be it =)

**Comment by ridiculousfish** [comment]
It's very surprising that the escape would be delivered separate from the other characters.

**Comment by TTimo** [comment]
Looking forward to attempt multi tab with fish shell again once this is fixed .. sticking to tmux for now .. any ETA?

**Comment by doxxx** [comment]
Windows 7 74-bit, Cygwin 64-bit, fish installed from Cygwin repo, mintty (not inside ConEmu).

I'm getting some weird behaviour from the Escape key: Pressing the Escape key just inserts an unknown character (for the font I'm using, Consolas, it's a question mark in a box).

Even though LANG is already set to en_US.UTF-8 (probably by Cygwin from my Windows locale settings), if I set LANG to en_US.UTF-8 in fish using `set -Ux LANG en_US.UTF-8` (which gives a warning `set: Warning: universal scope selected, but a global variable “LANG” exists.`) and restart mintty+fish, pressing the Escape key causes the the cursor to move once to the left. Pressing it again moves it once to the right. Pressing it again doesn't do anything. However, after any number of presses, pressing any other key results in "Wide character 65535 has no narrow representation" to be spammed in the terminal window. After this, the terminal seems to be a bit broken because typing anything will randomly spam that message.

**Comment by alkemir** [comment]
I think it has to do with Windows breaking the characters corresponding to ESC key into two meesages. Please read the above discussion. Probably having a larger WAIT_ON_ESCAPE can alliviate partially the situation. It is interesting that two different console emulators will behave different; as it will allow fish to pinpoint the place in the workflow where character delivery gets broken.

**Comment by doxxx** [comment]
I just experimented with `xxd` to see what byte sequences were produced by various keys. As far as I can tell, Escape on its own is producing 0x1b. I determined this by starting xxd and then typing Escape, Enter, Ctrl-D, and `xxd` outputs `1b0a`.

I found key_reader.c in the fish repo but I haven't been able to find the corresponding executable in the cygwin fish installation. I'm going to try downloading the fish source and compiling it.

**Comment by doxxx** [comment]
I noticed these warnings during the compile:

```
src/input_common.cpp: In function ‘wchar_t input_common_readch(int)’:
src/input_common.cpp:237:28: warning: large integer implicitly truncated to unsigned type [-Woverflow]
                     return WEOF;
                            ^
src/input_common.cpp:240:28: warning: large integer implicitly truncated to unsigned type [-Woverflow]
                     return WEOF;
                            ^
```

That might explain why I'm seeing the error about 65535? WEOF == -1, which when converted to unsigned integer becomes 65535.

**Comment by doxxx** [comment]
I got key_reader compiled from the source, and it outputs `dec: 27 hex: 1b` when I press just the Escape key.

**Comment by doxxx** [comment]
Something else which I think is related to this issue, is that `make test` has a few failures:

```
Testing error messages
Error: Should have been able to unescape \U10FFFF
```

And:

```
Testing wide/narrow string conversion
Error: Line 387 - 21
Error: Line 387 - 82
<snip many similar lines>
Error: Line 387 - 3062: Conversion cycle of string MYள produced different string MY
<snip>
Error: Line 387 - 18679: Conversion cycle of string >﮳ produced different string >
<snip>
Error: Line 387 - 44086: Conversion cycle of string 뾳 produced different string
<snip>
Error: Line 387 - 46830: Conversion cycle of string 䞳 produced different string
<snip>
Error: Line 387 - 88107: Conversion cycle of string  produced different string
<snip>
Error: Line 387 - 98647: Conversion cycle of string     ﾳ produced different string
```

The tests subsequently gets stuck on `Testing iothreads`, which is probably a separate issue.

**Comment by doxxx** [comment]
Running the tests again gets past `Testing iothreads` and fails in `Testing utils`:

```
Error: u2w: resulted in forbitten wchars (len): FAILED (rv: 3, must be 0)
Error: u2w: calculate length: FAILED (rv: 6, must be 0)
Error: u2w: calculate length, ignore bad chars: FAILED (rv: 2, must be 0)
Error: w2u: 4 octets chars: FAILED (rv: 7, must be 9)
Error: w2u: 5 octets chars: FAILED (rv: 9, must be 16)
Error: w2u: 6 octets chars: FAILED (rv: 9, must be 19)
Error: w2u: ignore bad chars: FAILED (rv: 0, must be 3)
Error: w2u: mixed languages: FAILED (rv: 0, must be 18)
Error: w2u: boundaries +1: FAILED (rv: 0, must be 18)
Error: w2u: calculate length: FAILED (rv: 0, must be 18)
Error: w2u: calculate length of bad chars: FAILED (rv: 9, must be 0)
Error: w2u: calculate length, ignore bad chars: FAILED (rv: 9, must be 3)
```

Which also seem related.

**Comment by BretJohnson** [comment]
What's the latest status here?   Any tips so I can get fish + cygwin/ConEmu to work with arrow keys?  It'd be great to have some of workaround or solution here.  Thanks.

**Comment by faho** [comment]
@BretJohnson: We now (since 2646d51a) have a configurable escape timeout via the fish_escape_delay_ms variable which might be useful here. If you do test it, please report back!

**Comment by krader1961** [comment]
I cannot reproduce this. There have been a lot of relevant changes in the past year and probably many more before then. So I'm closing this issue. If anyone can reproduce this using fish built from source as of this date please reopen the issue.

**Comment by tkkcc** [comment]
I have this issue when using `qsub` to enter nodes
```sh
qsub -I -l nodes=node1:ppn=2:gpus=1 -N debug -S ~/.linuxbrew/bin/fish -X -d .
```
after set `fish_escape_delay_ms` from 10 to 100, it's gone
